psi4/src/psi4/libfock/composite.cc

```diff
@@ -0,0 +1,794 @@
+    }
+
+    // Temporary buffers for J to minimize race tonditions
+    std::vector<std::vector<SharedMatrix>> JT;
+
+    for (int thread = 0; thread < nthread_; thread++) {
+        std::vector<SharedVector> gp2;
+        std::vector<SharedMatrix> J2;
+        for (size_t ind = 0; ind < nmat; ind++) {
+            J2.push_back(std::make_shared<Matrix>(max_nbf_per_shell, max_nbf_per_shell));
+            J2[ind]->zero();
+        }
+        JT.push_back(J2);
+    }
+
+    // Solve for gammaP = (P|rs)*Drs
+#pragma omp parallel for num_threads(nthread_) schedule(dynamic)
+    for (int P = 0; P < aux_nshell; P++) {
+
+        int p_start = auxiliary_->shell_to_basis_function(P);
+        int num_p = auxiliary_->shell(P).nfunction();
+
+        int thread = 0;
+#ifdef _OPENMP
+        thread = omp_get_thread_num();
+#endif
+
+        for (const auto& UV : shell_pairs) {
+
+            int U = UV.first;
+            int V = UV.second;
+
+            int u_start = primary_->shell_to_basis_function(U);
+            int num_u = primary_->shell(U).nfunction();
+
+            int v_start = primary_->shell_to_basis_function(V);
+            int num_v = primary_->shell(V).nfunction();
+
+            ints_[thread]->compute_shell(P, 0, U, V);
+
+            const double* buffer = ints_[thread]->buffer();
+
+            double prefactor = 2.0;
+            if (U == V) prefactor *= 0.5;
+
+            for (int i = 0; i < D.size(); i++) {
+                double** Dp = D[i]->pointer();
+                double *Puv = const_cast<double *>(buffer);
+
+                /*
+                for (int p = p_start; p < p_start + num_p; p++) {
+                    int dp = p - p_start;
+                    for (int u = u_start; u < u_start + num_u; u++) {
+                        int du = u - u_start;
+                        for (int v = v_start; v < v_start + num_v; v++) {
+                            int dv = v - v_start;
+                            gamp[i * naux + p] += prefactor * (*Puv) * Dp[u][v];
+                            Puv++;
+                        }
+                    }
+                }
+                */
+
+                std::vector<double> Dbuff(num_u * num_v, 0.0);
+                for (int u = u_start; u < u_start + num_u; u++) {
+                    int du = u - u_start;
+                    for (int v = v_start; v < v_start + num_v; v++) {
+                        int dv = v - v_start;
+                        Dbuff[du * num_v + dv] = Dp[u][v];
+                    }
+                }
+                C_DGEMV('N', num_p, num_u * num_v, prefactor, (double *)Puv, num_u * num_v, Dbuff.data(), 1, 1.0, &(gamp[i * naux + p_start]), 1);
+            }
+        }
+    }
+
+    // Solve for gammaQ, (P|Q) * gammaQ = gammaP
+    SharedMatrix Jmet_copy = Jmet_->clone();
+
+    std::vector<int> ipiv(naux);
+    C_DGESV(naux, nmat, Jmet_copy->pointer()[0], naux, ipiv.data(), gamp, naux);
+
+#pragma omp parallel for num_threads(nthread_) schedule(dynamic)
```

Comment:
Okay, second discussion at this point. I read your comment on my initial comments on parallelism, and after looking at the code, I agree - it is very important to avoid race conditions and unnecessary thread synchronization. However, my second point stands - you can probably gain parallel efficiency by doing some loop fusions. 

From a first glance, for this second loop, you could loop over the combined shell pair index UV instead of simply the shell U. This would make the granularity of the parallelism of this second loop smaller, improving load balance. Additionally, I have a hunch that doing this wouldn't decrease the granularity too much, because each UV loops over the entire set of auxiliary basis set shells. If looping over UV does decrease granularity too much, then the solution is simple - increase the OpenMP batch size. Doing a loop fusion as such would also come with the benefit of eliminating the need to construct shell_partners at all - a small benefit, but it's something.

Reply:
I agree, the `shell_partners` and separate loops were an artifact from debugging, and it can go away safely